packages/abi-coder/src/fragments/param-type.ts

```diff
@@ -0,0 +1,175 @@
+  readonly components?: ReadonlyArray<JsonFragmentType>;
+  readonly typeArguments?: ReadonlyArray<JsonFragmentType>;
+}
+
+const CONSTRUCTOR_GUARD = {};
+
+function populate(object: ParamType, params: { type?: string } & ParamTypeProps) {
+  Object.keys(params).forEach((key) => {
+    const paramTypeKey = key as keyof ParamTypeProps;
+    const value = params[paramTypeKey];
+
+    defineReadOnly(object, paramTypeKey, value);
+  });
+}
+
+const paramTypeArray = /^(.*)\[([0-9]*)\]$/;
+
+export interface ParamTypeProps {
+  // The local name of the parameter (of null if unbound)
+  readonly name?: string;
+
+  // The fully qualified type (e.g. "address", "tuple(address)", "uint256[3][]"
+  readonly type: string;
+
+  // The base type (e.g. "address", "tuple", "array")
+  readonly baseType?: string;
+
+  // Indexable Paramters ONLY (otherwise null)
+  readonly indexed?: boolean;
+
+  // Tuples ONLY: (otherwise null)
+  //  - sub-components
+  readonly components?: Array<ParamType>;
+
+  // Tuples ONLY: (otherwise null)
+  //  - sub-components
+  readonly typeArguments?: Array<ParamType>;
+
+  // Arrays ONLY: (otherwise null)
+  //  - length of the array (-1 for dynamic length)
+  //  - child type
+  readonly arrayLength?: number;
+  readonly arrayChildren?: ParamType;
+
+  readonly isParamType?: boolean;
+}
+
+export class ParamType implements ParamTypeProps {
+  readonly name?: string;
+  readonly type!: string;
+  readonly baseType?: string;
+  readonly indexed?: boolean;
+  readonly components?: Array<ParamType>;
+  readonly typeArguments?: Array<ParamType>;
+  readonly arrayLength?: number;
+  readonly arrayChildren?: ParamType;
+  readonly isParamType?: boolean;
+
+  constructor(constructorGuard: typeof CONSTRUCTOR_GUARD, params: ParamTypeProps) {
+    if (constructorGuard !== CONSTRUCTOR_GUARD) {
+      abiLogger.throwError('use fromString', Logger.errors.UNSUPPORTED_OPERATION, {
+        operation: 'new ParamType()',
+      });
+    }
+    populate(this, params);
+
+    const match = this.type.match(paramTypeArray);
+    if (match && match[1]) {
+      populate(this, {
+        arrayLength: parseInt(match[2] || '-1', 10),
+        arrayChildren: ParamType.fromObject({
+          type: match[1],
+          components: this.components,
+          typeArguments: this.typeArguments,
+        }),
+        baseType: 'array',
+        type: this.type,
+      });
+    } else {
+      populate(this, {
+        arrayLength: undefined,
+        arrayChildren: undefined,
+        baseType: this.components !== undefined ? 'tuple' : this.type,
+        type: this.type,
+      });
+    }
+
+    this.isParamType = true;
+
+    Object.freeze(this);
+  }
+
+  getSignaturePrefix(): string {
+    if (this.type) {
+      const arrayMatch = arrayRegEx.exec(this.type)?.groups;
+      const structMatch = structRegEx.exec(this.type)?.groups;
+      const enumMatch = enumRegEx.exec(this.type)?.groups;
+
+      // TODO: return 'a' after this PR is merged
```

Comment:
Just a reminder to add more details in the comments. To reduce required context knowledge.